### PR TITLE
runnable: reintroduce into_raw/from_raw functions.

### PR DESCRIPTION
**Pull request by osiewicz**
These APIs were removed prior to async-task 4.0 (in commit deb709f4198c4c6d1c691a69f5cd713d040e0fed); they can however be useful to e.g. pass Runnable as a context argument to GCD's [dispatch_async_f](https://developer.apple.com/documentation/dispatch/1452834-dispatch_async_f), which takes an opaque `NonNull<()>` as an argument. Without exposing from_raw/into_raw one has to box the Runnable in order to get something that can be passed across that boundary, which seems wasteful.

### CODE REVIEW

File: src/runnable.rs [hunk 1]
```diff
@@ -820,6 +820,26 @@ impl<M> Runnable<M> {
     fn header(&self) -> &Header<M> {
         unsafe { &*(self.ptr.as_ptr() as *const Header<M>) }
     }
+
+    /// Converts this task into a raw pointer.
+    pub fn into_raw(self) -> NonNull<()> {
+        let ptr = self.ptr;
+        mem::forget(self);
+        ptr
+    }
```
Comment (notgull) [inline src/runnable.rs]: Add an example to this function to demonstrate its use.

Comment (notgull) [reply src/runnable.rs]: Also, a more thorough documentation of the guarantees (or lack thereof) affording by this function would be nice. The pointer is non-null, but are there any guarantees as to what it will dereference to, or what the state of the task will be?

File: src/runnable.rs [hunk 2]
```diff
@@ -820,6 +820,26 @@ impl<M> Runnable<M> {
...
+
+    /// Converts a raw pointer into a task.
+    ///
+    /// # Safety
+    /// This method should only be used with raw pointers returned from [`into_raw`].
+    ///
+    /// [`into_raw`]: #method.into_raw
```
Comment (notgull) [inline src/runnable.rs]: Ditto here.

Also, there should be a newline under `# Safety`.

File: src/runnable.rs [hunk 3]
```diff
@@ -820,6 +820,26 @@ impl<M> Runnable<M> {
...
+    pub unsafe fn from_raw(ptr: NonNull<()>) -> Self {
```
Comment (notgull) [inline src/runnable.rs:885]: Should we use this function to construct the initial `Runnable`?

File: tests/basic.rs [hunk 4]
```diff
@@ -297,3 +297,12 @@ fn waker() {
...
 }
+
+#[test]
+fn raw() {
+    let (runnable, _handle) = async_task::spawn(async {}, |_| panic!());
+
+    let a = runnable.into_raw();
+    let task = unsafe { Runnable::<()>::from_raw(a) };
+    task.run();
+}
```
Comment (notgull) [inline tests/basic.rs]: I would like a more thorough test that passes the task across a simulated FFI boundary. Like, have an `extern "C"` function that takes a pointer, just to see if this works for its intended use case.

**Review by notgull — CHANGES_REQUESTED** [review]
I'm not entirely sure why Stjepan originally removed the `raw` methods. @yoshuawuyts  Do you have any insight? I think you were involved in the original design of this task system.

File: src/runnable.rs (hunk 1 above)
Comment (osiewicz) [reply src/runnable.rs]: I think we can't really give any guarantees as the task might be cancelled in the meantime, e.g:
```rust
    #[test]
    fn cancelation_can_change_the_state_of_the_task() {
        let (runnable, task) = crate::spawn(async { 1 + 2 }, |_| {});
        let expected_state =
            (crate::state::TASK | crate::state::SCHEDULED) + crate::state::REFERENCE;
        assert_eq!(
            runnable
                .header()
                .state
                .load(core::sync::atomic::Ordering::Relaxed),
            expected_state
        );
        let raw = runnable.into_raw();
        let _ = task.cancel();
        let runnable = unsafe { Runnable::<()>::from_raw(raw) };
        assert_ne!(
            runnable
                .header()
                .state
                .load(core::sync::atomic::Ordering::Relaxed),
            expected_state
        );
    }
```
`are there any guarantees as to what it will dereference to` can you elaborate on that question? Should the end user care about what the underlying object is?
On the other hand, there is one guarantee that we can make - the pointer should be safe to pass into `from_raw` at any time, since we're reference counting under the hood. 
On yet another hand, the user **should not** be reusing that pointer - any pointer returned from `into_raw` is safe to pass into `from_raw` at most once (or should I say `exactly once`).

**Review by notgull — APPROVED** [review]
Thanks!
